### PR TITLE
Production ready deployment pipeline

### PR DESCRIPTION
**Pull request by merrilmathew**
### Context

Add production ready deployment pipeline with terraform support. This PR will deploy the app and backing services using Teacher services standards and follows other TRA deployments.

### Trello card

https://trello.com/c/Uxm2n5Zh

### CODE REVIEW

**Comment by merrilmathew** [comment]
[Production-ready deployment pipeline and DevOps](https://trello.com/c/Uxm2n5Zh/121-production-ready-deployment-pipeline-and-devops)

File: Dockerfile [hunk 1]
```diff
@@ -2,18 +2,23 @@ FROM ruby:3.0.3-alpine
...
+
+COPY package.json /myapp/
+COPY yarn.lock /myapp/
+RUN yarn install --frozen-lockfile
+
 ADD . /myapp
+ENV RAILS_ENV=production \
+    RAILS_SERVE_STATIC_FILES=yes \
+    LANG=en_GB.UTF-8 \
+    SECRET_KEY_BASE=TestKey
```
Comment (tvararu) [inline Dockerfile:19]: Here we're generating the docker image using this temporary key, but when we deploy later, are we using a different `SECRET_KEY_BASE` in the application environment?

---

[More background on what SECRET_KEY_BASE does](https://stackoverflow.com/questions/25426940/what-is-the-use-of-secret-key-base-in-rails-4/25427302#25427302).

TL;DR it can be anything, including a randomly generated GUID that we don't have to know / retrieve ever. It shouldn't live in an encrypted storage like KeyVault because we should only write it once and then forget about it.

When it gets replaced, all session cookies become invalid, so it shouldn't be changed between routine deployment (it should be persistent).

It should also be different between all envs.

Comment (merrilmathew) [reply Dockerfile:19]: used terraform random_uuid to pass in env variable on deployment

File: terraform/app.tf [hunk 2]
```diff
@@ -0,0 +1,44 @@
+resource "random_uuid" "id" {
+}
+locals {
+  secret_key_base = {
+    SECRET_KEY_BASE = "${random_uuid.id.result}"
```
Comment (tvararu) [inline terraform/app.tf:5]: Just to check: this won't change SECRET_KEY_BASE between deployments right? Just once on the initial deploy.

**Review by tvararu — APPROVED** [review]
Can be merged as-is and then we can write another PR to fix the `test.yml` and split it up so that tests still run on branches/PRs.

File: terraform/app.tf (hunk 2 above)
Comment (merrilmathew) [reply terraform/app.tf:5]: it will change each time it deploys. Do you want one constant value for each environment?

Comment (tvararu) [reply terraform/app.tf:5]: Yeah. We need a constant value, because otherwise cookies will stop working between each deploy, and log users out.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -1,17 +1,35 @@
...
+    inputs:
+      environment:
+        description: 'Deploy environment ( dev, preprod or production )'
+        required: true
+        default: dev
+        type: environment
   push:
     branches:
-      - '**'
+      - main
```
[the remark this answers is not in the document]
Comment (merrilmathew) [reply .github/workflows/test.yml:13]: ok. Let me try to fix that.

**Review by tvararu — APPROVED** [review]
